### ISSUE TITLE
F821: false positive when using `del` in `.pyi` files

### ISSUE THREAD

**Issue by loic-simon** [astral-sh/ruff#23539]
### Summary

Consider the following 

```py
# mre.pyi
class MyClass: ...
def f(cls: MyClass) -> None: ...
del MyClass
```

```bash
$ ruff check --isolated mre.pyi
F821 Undefined name `MyClass`
 --> mre.pyi:2:12
  |
1 | class MyClass: ...
2 | def f(cls: MyClass) -> None: ...
  |            ^^^^^^^
3 | del MyClass
  |

Found 1 error.
```

If I rename the file `mre.py` instead of `.pyi`, the check pass ([playground](https://play.ruff.rs/27bfcaf3-2b51-441a-a14f-d581f48262f7)).

AFAIK, `del` statements are allowed in `.pyi` files _(and are useful to avoid exposing symbols needed for typing but not exposed at runtime, but maybe there is a better way to do that?)_

Possibly related to 
  * https://github.com/astral-sh/ruff/issues/3011 - myby even caused by its fix? 
  * https://github.com/astral-sh/ruff/issues/6242?

### Version

ruff 0.15.2 (9d18ee911 2026-02-19)

**Comment by ntBre** [comment]
Thanks for the report! @AlexWaygood may be able to correct me, but from searching around a bit, I don't think using `del` like this is the most common way to avoid exporting names. It seems like either using `__all__`, prefixing  private names with `_`, or decorating with `typing.type_check_only` are more common, based on my reading of https://typing.python.org/en/latest/guides/writing_stubs.html#stub-only-objects.

This still seems like a bug since it works as expected in a `.py` file, but there may be other alternatives in stub files too.

**Comment by AlexWaygood** [comment]
> Thanks for the report! [@AlexWaygood](https://github.com/AlexWaygood) may be able to correct me, but from searching around a bit, I don't think using `del` like this is the most common way to avoid exporting names. It seems like either using `__all__`, prefixing private names with `_`, or decorating with `typing.type_check_only` are more common, based on my reading of [typing.python.org/en/latest/guides/writing_stubs.html#stub-only-objects](https://typing.python.org/en/latest/guides/writing_stubs.html#stub-only-objects).

Yes, I think that's all correct. I don't think using `del` in a stub file is _unsupported_ (I don't think there's anything anywhere saying that it's banned) -- but I also think it's pretty unusual to use `del` in a stub file (which is generally meant to be contain declarative statements rather than imperative logic that would depend on control flow), and arguably not very idiomatic.

**Comment by stakeswky** [comment]
This should be fixed by #23550, which suppresses F821 for names deleted via `del` in stub files (`.pyi`).
